src/hotspot/share/opto/loopopts.cpp

```diff
@@ -1445,7 +1445,8 @@ void PhaseIdealLoop::try_sink_out_of_loop(Node* n) {
       !n->is_MergeMem() &&
       !n->is_CMove() &&
       !is_raw_to_oop_cast && // don't extend live ranges of raw oops
-      n->Opcode() != Op_Opaque4) {
+      n->Opcode() != Op_Opaque4 &&
+      (n->Opcode() == Op_CastII && ((CastIINode*)n)->has_range_check())) {
```

Comment:
Right! Thanks for catching this. Actually I changed the fix to prevent all nodes that can capture a type from being sunk for now to be on the safe side.

Reply:
Sounds good, I guess you will undo that with your general fix later?

Reply:
That's the plan.